File: backend/app/models/knowledge.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.core.db import Base


class KnowledgeEntry(Base):
    __tablename__ = "knowledge_entries"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    category: Mapped[str] = mapped_column(String(32), index=True)  # course|faq|tone|banned|staff|process
    title: Mapped[str] = mapped_column(String(255), default="")
    content: Mapped[str] = mapped_column(Text, default="")
    tags: Mapped[str] = mapped_column(String(255), default="")
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    updated_by: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True, index=True)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
    )

File: frontend/src/api/knowledge.ts
import client from './client'

export type KnowledgeCategory = 'course' | 'faq' | 'tone' | 'banned' | 'staff' | 'process'

export interface KnowledgeEntry {
  id: number
  category: string
  title: string
  content: string
  tags: string
  is_active: boolean
  updated_by?: number | null
  updated_at?: string | null
}

export interface KnowledgeCreateInput {
  category: KnowledgeCategory | string
  title?: string
  content?: string
  tags?: string
  is_active?: boolean
}

export interface KnowledgeUpdateInput {
  category?: KnowledgeCategory | string
  title?: string
  content?: string
  tags?: string
  is_active?: boolean
}

interface ApiResponse<T> {
  data: T | null
  error: { code: string; message: string } | null
}

function unwrap<T>(res: { data: ApiResponse<T> }, fallback: string): T {
  if (!res.data.data) {
    throw new Error(res.data.error?.message || fallback)
  }
  return res.data.data
}

export async function listKnowledge(category?: string): Promise<KnowledgeEntry[]> {
  const res = await client.get<ApiResponse<KnowledgeEntry[]>>('/knowledge', {
    params: category ? { category } : undefined,
  })
  return unwrap(res, 'Failed to list knowledge')
}

export async function createKnowledge(payload: KnowledgeCreateInput): Promise<KnowledgeEntry> {
  const res = await client.post<ApiResponse<KnowledgeEntry>>('/knowledge', payload)
  return unwrap(res, 'Failed to create knowledge')
}

export async function updateKnowledge(
  id: number,
  payload: KnowledgeUpdateInput,
): Promise<KnowledgeEntry> {
  const res = await client.patch<ApiResponse<KnowledgeEntry>>(`/knowledge/${id}`, payload)
  return unwrap(res, 'Failed to update knowledge')
}

export async function deleteKnowledge(id: number): Promise<{ id: number }> {
  const res = await client.delete<ApiResponse<{ id: number }>>(`/knowledge/${id}`)
  return unwrap(res, 'Failed to delete knowledge')
}
